lib/shopify_api/api_version.rb

```diff
@@ -3,111 +3,143 @@ module ShopifyAPI
     class InvalidVersion < StandardError; end
+    include Comparable
 
-    extend DefinedVersions
+    HANDLE_FORMAT = /^\d{4}-\d{2}$/.freeze
+    UNSTABLE_HANDLE = 'unstable'
+    UNSTABLE_AS_DATE = Time.utc(3000, 1, 1)
+    API_PREFIX = '/admin/api/'
+    COERSION_MODES = [:predefined_only, :define_on_unknown].freeze
 
-    include Comparable
+    class << self
+      attr_reader :versions
 
-    def self.coerce_to_version(version_or_name)
-      return version_or_name if version_or_name.is_a?(ApiVersion)
+      def coercion_mode
+        @coercion_mode ||= :define_on_unknown
+      end
 
-      @versions ||= {}
-      @versions.fetch(version_or_name.to_s) do
-        raise UnknownVersion, "#{version_or_name} is not in the defined version set: #{@versions.keys.join(', ')}"
+      def coercion_mode=(mode)
+        raise ArgumentError, "Mode must be one of #{COERSION_MODES}" unless COERSION_MODES.include?(mode)
+        sanitize_known_versions if mode == :predefined_only
+        @coercion_mode = mode
+      end
+
+      def coerce_to_version(version_or_handle)
+        return version_or_handle if version_or_handle.is_a?(ApiVersion)
+        handle = version_or_handle.to_s
+
+        @versions ||= {}
+        @versions.fetch(handle) do
+          if @coercion_mode == :predefined_only
+            error_msg = if @versions.empty?
+              "No versions defined. You must call `ApiVersion.define_known_versions` first."
+            else
+              "`#{handle}` is not in the defined version set. Available versions: #{@versions.keys}"
+            end
+            raise UnknownVersion, "ApiVersion.coercion_mode is set to `:predefined_only`. #{error_msg}"
+          else
+            @versions[handle] = ApiVersion.new(handle: handle)
+          end
+        end
       end
-    end
 
-    def self.define_version(version)
-      @versions ||= {}
+      def define_known_versions
+        @versions = Meta.admin_versions.map { |version| [version.handle, ApiVersion.new(version.attributes, true)] }.to_h
+      end
 
-      @versions[version.name] = version
-    end
+      def clear_defined_versions
+        @versions = {}
+      end
 
-    def self.clear_defined_versions
-      @versions = {}
+      def latest_stable_version
+        warn(
+          '[DEPRECATED] ShopifyAPI::ApiVersion.latest_stable_version is deprecated and will be removed in a future version.'
+        )
+        versions.values.find(&:latest_supported?)
+      end
+
+      private
+
+      def sanitize_known_versions
+        return if @versions.nil?
+        @versions = @versions.keys.map do |handle|
+          next unless @versions[handle].verified?
+          [handle, @versions[handle]]
+        end.compact.to_h
+      end
     end
 
-    def self.latest_stable_version
-      @versions.values.select(&:stable?).sort.last
+    attr_reader :handle, :display_name, :supported, :latest_supported, :verified
+
+    def initialize(attributes = {}, verified = false)
```

Comment:
Is there any reason to give these two default values? They're only constructed in two places and I'd think explicit would be better for both.

Reply:
ehh, theres no real reason to have attributes defaulted beyond it being a a hold over from converting from active resource. If it defaults to `{}` it will raise an error. Verified I do think should have a default though since it allows us to define a version with `ApiVersion.new('2019-10')` and automatically know its not verified. When we call `define_known_versions` we explicitly set verified to true because its from the server.